### PR TITLE
BUG: Order percentile monotonically

### LINKED ISSUE #14685: BUG: numpy.percentile output is not sorted
The output of `numpy.percentile` is not always sorted

### Reproducing code example:

```python
import numpy as np
q = np.arange(0, 1, 0.01) * 100
percentile = np.percentile(np.array([0, 1, 1, 2, 2, 3, 3 , 4, 5, 5, 1, 1, 9, 9 ,9, 8, 8, 7]) * 0.1, q)
equals_sorted = np.sort(percentile) == percentile
print(equals_sorted)
assert equals_sorted.all()
```

### Error message:
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True False False  True  True  True  True False
  True  True  True False]
AssertionError                            Traceback (most recent call last)
<ipython-input-63-2850fe4a4ce3> in <module>
      1 q = np.percentile(np.array([0, 1, 1, 2, 2, 3, 3 , 4, 5, 5, 1, 1, 9, 9 ,9, 8, 8, 7]) * 0.1, np.arange(0, 1, 0.01) * 100)
      2 equals_sorted = np.sort(q) == q
----> 3 assert equals_sorted.all()

AssertionError: 

### Numpy/Python version information:

1.17.2 3.6.8 (v3.6.8:3c6b436a57, Dec 24 2018, 02:04:31) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]

### PR DESCRIPTION
**Pull request by CloseChoice**
fixes #14685

Order output of `np.quantile`/`np.percentile` monotonically if quantiles/percentiles are ordered monotonically.

Change the way of performing the linear interpolation from (2) to (1) as given in 
[https://math.stackexchange.com/questions/907327/accurate-floating-point-linear-interpolation](https://math.stackexchange.com/questions/907327/accurate-floating-point-linear-interpolation)

### CODE REVIEW

**Comment by eric-wieser** [comment]
This is a duplicate of #15098. For reviewers:

* `master` uses `lerp(a, b, t) = a*(1-t) + b*t`
* This Pr uses `lerp(a, b, t) = a + (b-a)*t`
* #15098 uses `lerp(a, b, t)  = a + (b-a)*t if t < 0.5 else b - (b-a)*(1-t)`

**Comment by CloseChoice** [comment]
> This is a duplicate of #15098. For reviewers:
> 
>     * `master` uses `lerp(a, b, t) = a*(1-t) + b*t`
> 
>     * This Pr uses `lerp(a, b, t) = a + (b-a)*t`
> 
>     * #15098 uses `lerp(a, b, t)  = a + (b-a)*t if t < 0.5 else b - (b-a)*(1-t)`

Hi,
thanks for the answer and you're totally correct concerning my and master's formula. I just realized that there is an open PR when I was done writing my code, so I thought I'd try to give it a chance. Can close this PR though, if you advice to.

File: numpy/lib/tests/test_function_base.py
Comment (CloseChoice) [inline numpy/lib/tests/test_function_base.py:3098]: I actually thought of writing this as a hypothesis test but hypothesis module was never imported and I don't know your conventions. But can change it hypothesis for sure if that is desired.

File: numpy/lib/function_base.py
Comment (CloseChoice) [inline numpy/lib/function_base.py:3979]: I did not understand what `out` means in this context. If you can give me a hint I might be able to simplify the code even more. I guess there is the possibility to get rid of `x1 = x1.squeeze(0)`

Comment (eric-wieser) [reply numpy/lib/function_base.py:3979]: In working out what hint to give you, I ended up doing some cleanup of this code myself. Thanks for drawing it to my attention. It might be best for you to wait until my cleanup goes through, and then to rebase your tests.

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [reply numpy/lib/tests/test_function_base.py:3098]: We recently decided we were willing to accept hypothesis tests - if you think the test is easier to read with hypothesis, please use it, so that maintainers like me who haven't used it can get the hang of it.

**Comment by eric-wieser** [comment]
Thanks for your patience - my cleanup is in. It's unlikely you'll be able to usefully merge the implementation, but you should be able to keep the tests.

File: numpy/lib/tests/test_function_base.py
Comment (CloseChoice) [reply numpy/lib/tests/test_function_base.py]: I added a hypothesis test. Am curious for your feedback. The purpose of this test is not to increase readability but to cover more test cases without writing multiple tests and to find failing egde cases we might not have thought of.

File: numpy/lib/function_base.py
Comment (CloseChoice) [reply numpy/lib/function_base.py]: > Thanks for your patience - my cleanup is in. It's unlikely you'll be able to usefully merge the implementation, but you should be able to keep the tests.

Your cleanup made it way easier to implement the montonic lerp. Thanks for that.

File: numpy/lib/function_base.py
Comment (eric-wieser) [inline numpy/lib/function_base.py:3948]:
```suggestion
```

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3116]:
```suggestion
        assert_equal(np.sort(quantile), quantile)
```

**Comment by eric-wieser** [comment]
My fear with this patch is that we also care about the other properties in https://github.com/numpy/numpy/issues/14685#issuecomment-54277321 (vs #15098 which claims to satisfy all of them).

Do you think you could put together a hypothesis tests that can prove that those properties are not satisfied with your solution?

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3147]:
```suggestion
        assert np.lib.function_base._lerp(a, b, t) == np.lib.function_base._lerp(b, a, 1-t)
```

Comment (eric-wieser) [reply numpy/lib/tests/test_function_base.py:3147]: Or more likely,

```suggestion
        assert np.lib.function_base._lerp(a, b, 1 - (1- t)) == np.lib.function_base._lerp(b, a, 1 - t)
```
To ensure that the precision of t is the same in both directions

**Comment by CloseChoice** [comment]
> My fear with this patch is that we also care about the other properties in [#14685 (comment)](https://github.com/numpy/numpy/issues/14685#issuecomment-54277321) (vs #15098 which claims to satisfy all of them).
> 
> Do you think you could put together a hypothesis tests that can prove that those properties are not satisfied with your solution?

I added the tests but both solutions (mine and #15098) don't satisfy the exact symmetry condition but of course both satisfy the condition with `np.isclose`.

File: numpy/lib/tests/test_function_base.py
Comment (CloseChoice) [reply numpy/lib/tests/test_function_base.py]: done.

**Comment by eric-wieser** [comment]
Even with https://github.com/numpy/numpy/pull/16273#discussion_r430647408?

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3147]:
```suggestion
        # double subtraction is needed to remove the extra precision that t < 0.5 has
        assert np.lib.function_base._lerp(a, b, 1 - (1 - t)) == np.lib.function_base._lerp(b, a, 1 - t)
```

**Comment by CloseChoice** [comment]
> Even with [#16273 (comment)](https://github.com/numpy/numpy/pull/16273#discussion_r430647408)?

Only with [#16273 (comment)](https://github.com/numpy/numpy/pull/16273#discussion_r430647408). Using assert `np.lib.function_base._lerp(a, b, t) == np.lib.function_base._lerp(b, a, (1 - t))` fails it.

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3136]: These need to be 64 for this test to be interesting.

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3139]: I'm a bit worried this test is unsafe because these multiplications could cancel to 0 for the extreme cases that we're worried about. Can you write it as the safer:
```suggestion
        l0 = np.lib.function_base._lerp(a, b, t0)
        l1 = np.lib.function_base._lerp(a, b, t1)
        if t0 == t1 or a == b:
            assert l0 == l1  # uninteresting
        elif (t0 < t1) == (a < b):
            assert l0 < l1
        else:
            assert l0 > l1
```

File: numpy/lib/tests/test_function_base.py
Comment (CloseChoice) [reply numpy/lib/tests/test_function_base.py:3136]: that leads to overflows:

```
b=1.7976931348623157e+308
a=-1.7976931348623157e+308
np.subtract(b, a)
*** RuntimeWarning: overflow encountered in subtract
```

Comment (eric-wieser) [reply numpy/lib/tests/test_function_base.py:3136]: That's potentially a problem, the previous implementation did not fail like that for those cases.

For now, can you do something like clip a and b to reasonable magnitudes? We still want to test the precision of 64-bit, even if we decide we don't care about the range.

File: numpy/lib/function_base.py
Comment (eric-wieser) [reply numpy/lib/function_base.py:3881]: Do you need this path for anything other than the tests?

Comment (CloseChoice) [reply numpy/lib/function_base.py:3881]: Nope, only to ensure that the test output is array or scalar.

File: numpy/lib/tests/test_function_base.py
Comment (CloseChoice) [reply numpy/lib/tests/test_function_base.py]: I changed the range from `-1e300` up to `1e300`, seems to work fine.

Are you sure that it wasn't a problem before? The overflow happens in the basic `add` and subtract methods and should be reproducible for specific combinations for the former implementation aswell since we are still using those methods.

File: numpy/lib/function_base.py
Comment (CloseChoice) [reply numpy/lib/function_base.py]: You are correct. I added the 0d support. Am not sure about the lambda function though.

Comment (seberg) [reply numpy/lib/function_base.py]: @CloseChoice sorry, are we thinknig a bit differently here?  I do not think there is any possible call of `quantile` which requires correct support of 0D arrays.  If that is so, I think we should simply refuse to handle it here.  And note in the docstring that `_lerp` does not support 0D correctly.
The tests for `_lerp` can stay, simply make the inputs one dimensional arrays.

File: numpy/lib/function_base.py
Comment (eric-wieser) [inline numpy/lib/function_base.py:3889]: Shorter as:
```suggestion
    diff_b_a = subtract(b, a)
    # asanyarray is a stop-gap until gh-13105
    lerp_interpolation = asanyarray(add(a, diff_b_a*t, out=out))
    subtract(b, diff_b_a * (1 - t), out=lerp_interpolation, where=t>=0.5)
    if lerp_interpolation.ndim == 0 and out is None:
        lerp_interpolation = lerp_interpolation[()]  # unpack 0d arrays
    return lerp_interpolation
```
Most importantly, this means we don't have two separate code paths which both have to be tested

File: numpy/lib/function_base.py
Comment (CloseChoice) [reply numpy/lib/function_base.py]: @seberg thanks for the clarification. I just added @eric-wieser 's suggestion to use `asanyarray`. I guess this one needs go other all functions with extra handling of 0d as soon as a more generic solution is found. Is this a viable way to go with this PR?

Comment (eric-wieser) [reply numpy/lib/function_base.py]: > I do not think there is any possible call of quantile which requires correct support of 0D arrays.

Just to be clear @seberg, `quantile([1, 2, 3], 0.5)` is such a case where `lerp` is called with scalars.

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:3150]: style nit:
```suggestion
                      a=st.floats(allow_nan=False, allow_infinity=False,
                                  min_value=-1e300, max_value=1e300),
                      b=st.floats(allow_nan=False, allow_infinity=False,
                                  min_value=-1e300, max_value=1e300))
```

**Review by eric-wieser — COMMENTED** [review]
Looks good to me, assuming @seberg is happy.

File: numpy/lib/function_base.py
Comment (seberg) [reply numpy/lib/function_base.py]: Oh, ok, I thought since `axis=None` exists we always have 1-d effectively...

Comment (eric-wieser) [reply numpy/lib/function_base.py]: `axis` refers to the first argument, which is indeed always at least 1d. But `quantile(a, q).ndim == a.ndim + q.ndim - 1` - so when `q.ndim == 0`, the result of `lerp` is also 0d.

**Review by eric-wieser — APPROVED** [review]

**Review by seberg — APPROVED** [review]
LGTM, may just merge soon.  I cannot say, I love the way the indentation works out for the hypothesis tests. I think I would prefer a hanging indent, maybe of 8 spaces.  There may be long lines.

I also wonder if those tests (except the symmetric one maybe?), are actually likely to stress the issue. On first sight, it feels like we may need an absurd amount of fuzzy trials until the test could proof that the old version was bad?

EDIT: To be clear, I am not willing to stress on code style in the tests here...

**Comment by seberg** [comment]
Thanks @CloseChoice, lets put put this in :), I think the tests should cover the relevant paths well, even if I am not sure that all the tests actually stress the monotonicity very strongly.

**Comment by CloseChoice** [comment]
@seberg : Thanks for merging ;)

**Comment by glemaitre** [comment]
@CloseChoice @seberg @eric-wieser Thanks for taking care of this issue and sorry to have disappear from the map.
This will be really useful.

File: numpy/lib/function_base.py
Comment (HyukjinKwon) [inline numpy/lib/function_base.py:3878]: Just in case this was overlooked mistakenly, seems like this now disallows booleans in percentile:

```python
>>> np.percentile([True, False, False], q=0.5)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in percentile
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3818, in percentile
    return _quantile_unchecked(
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3937, in _quantile_unchecked
    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3515, in _ureduce
    r = func(a, **kwargs)
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 4064, in _quantile_ureduce_func
    r = _lerp(x_below, x_above, weights_above, out=out)
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3961, in _lerp
    diff_b_a = subtract(b, a)
TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
```

which worked before (NumPy 1.19):

```
0.0
```

<br/>
<br/>

FWIW, this behaviour change can also be triggered by pandas's:

```python
>>> pd.DataFrame({"i": [0, 1, 2], "b": [False, False, True], "s": ["x", "y", "z"]}).quantile(q=0.5, numeric_only=True)
```

Before:

```
i    1.0
b    0.0
Name: 0.5, dtype: float64
```

After:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../python3.8/site-packages/pandas/core/frame.py", line 9266, in quantile
    result = data._mgr.quantile(
  File "/.../python3.8/site-packages/pandas/core/internals/managers.py", line 491, in quantile
    block = b.quantile(axis=axis, qs=qs, interpolation=interpolation)
  File "/.../python3.8/site-packages/pandas/core/internals/blocks.py", line 1592, in quantile
    result = nanpercentile(
  File "/.../python3.8/site-packages/pandas/core/nanops.py", line 1675, in nanpercentile
    return np.percentile(values, q, axis=axis, interpolation=interpolation)
  File "<__array_function__ internals>", line 5, in percentile
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3818, in percentile
    return _quantile_unchecked(
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3937, in _quantile_unchecked
    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3515, in _ureduce
    r = func(a, **kwargs)
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 4064, in _quantile_ureduce_func
    r = _lerp(x_below, x_above, weights_above, out=out)
  File "/.../python3.8/site-packages/numpy/lib/function_base.py", line 3961, in _lerp
    diff_b_a = subtract(b, a)
TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
```

Comment (HyukjinKwon) [reply numpy/lib/function_base.py:3878]: The problem is that I am not sure if this is an intended behaviour or a bug. It could make sense either way 1. bool != numeric but 2. bool inherits int in Python.

I believe I provided enough details with a self-contained reproducer so I would defer to other maintainers here.

Comment (CloseChoice) [reply numpy/lib/function_base.py:3878]: Actually opening an issue with the regression tag would make sense here. What do you thi
nk @eric-wieser @seberg ?

Comment (seberg) [reply numpy/lib/function_base.py:3878]: Please open the issue, otherwise the discussion will get lost even quicker.  If we figure that it is OK if this breaks now, we can just close the issue again.  Thanks!

Comment (HyukjinKwon) [reply numpy/lib/function_base.py:3878]: done: https://github.com/numpy/numpy/issues/19154

**Pull request merged**
